File: backends/java/zappy-application/src/main/java/com/zappymart/application/cart/ChangeCartLineQuantity.java
package com.zappymart.application.cart;

import com.zappymart.application.Visitor;
import com.zappymart.application.ports.CartRepository;
import com.zappymart.application.ports.Clock;
import com.zappymart.application.ports.UnitOfWork;
import com.zappymart.domain.cart.Cart;
import com.zappymart.domain.shared.Result;

public final class ChangeCartLineQuantity {

    private final UnitOfWork unitOfWork;
    private final CurrentCart currentCart;
    private final CartRepository cartRepository;
    private final Clock clock;

    public ChangeCartLineQuantity(UnitOfWork unitOfWork, CurrentCart currentCart,
                                  CartRepository cartRepository, Clock clock) {
        this.unitOfWork = unitOfWork;
        this.currentCart = currentCart;
        this.cartRepository = cartRepository;
        this.clock = clock;
    }

    public Result<Cart> execute(Visitor visitor, String lineId, int quantity) {
        return unitOfWork.inTransaction(() -> {
            Cart cart = currentCart.forVisitor(visitor);
            return cart.changeLineQuantity(lineId, quantity, clock.now()).map(cartRepository::save);
        });
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/cart/ChangeCartLineQuantity.kt
package nl.zappymart.application.cart

import nl.zappymart.application.Visitor
import nl.zappymart.application.ports.CartRepository
import nl.zappymart.application.ports.Clock
import nl.zappymart.application.ports.UnitOfWork
import nl.zappymart.domain.shared.Result

class ChangeCartLineQuantity(
    private val visitorCart: VisitorCart,
    private val cartPromotion: CartPromotion,
    private val carts: CartRepository,
    private val clock: Clock,
    private val unitOfWork: UnitOfWork,
) {

    fun execute(visitor: Visitor, lineId: String, quantity: Int): CartChange = unitOfWork.execute {
        val cart = cartPromotion.refreshed(visitorCart.of(visitor))
        when (val changed = cart.withLineQuantityChanged(lineId, quantity, clock.moment())) {
            is Result.Success -> CartChange.made(carts.save(cartPromotion.refreshed(changed.value)))
            is Result.Refused -> CartChange.refused(cart, changed.errors, cart.lines.firstOrNull { line -> line.id == lineId }?.product)
        }
    }
}
